src/index.ts

```diff
@@ -148,30 +190,77 @@ export class CSSStyleObserver {
   }
 
   /**
-   * Collect CSS variable values and invoke callback.
+   * Process observed variables and collect changes
+   *
+   * @param computedStyle The computed style of the target element
+   * @param processChange Function to process each changed property
+   * @param finalizeChanges Function to finalize and handle collected changes
    */
-  private _handleUpdate(): void {
-    if (this._targetElement) {
-      const computedStyle = getComputedStyle(this._targetElement);
-
-      const changedProperties: CSSDeclarations = {};
-
-      this._observedVariables.forEach(propertyName => {
+  private _processChanges(
+    computedStyle: CSSStyleDeclaration,
+    processChange: (propertyName: string, currentValue: string, previousValue: string, hasChanged: boolean) => void,
+    finalizeChanges: () => void
+  ) {
+      this._observedVariables.forEach((propertyName) => {
         const currentValue = computedStyle.getPropertyValue(propertyName);
         const previousValue = this._cachedValues[propertyName] || '';
+        const hasChanged = currentValue !== previousValue;
+
+      if (this._notificationMode === NotificationMode.ALL || hasChanged) {
+        processChange(propertyName, currentValue, previousValue, hasChanged);
+        this._cachedValues[propertyName] = currentValue;
+      }
+    });
+
+    finalizeChanges();
+  }
 
-        if (this._notificationMode === NotificationMode.ALL || currentValue !== previousValue) {
-          changedProperties[propertyName] = currentValue;
-          this._cachedValues[propertyName] = currentValue;
+  /**
+   * Handlers for return formats
+   */
+  private _returnFormatHandlers: Record<ReturnFormat, (computedStyle: CSSStyleDeclaration) => void> = {
+    [ReturnFormat.VALUE_ONLY]: (computedStyle) => {
+      const changes: CSSDeclarations = {};
+
+      this._processChanges(computedStyle, (propertyName, currentValue) => {
+        changes[propertyName] = currentValue;
+      }, () => {
+        if (Object.keys(changes).length > 0) {
+          this._callback(changes);
         }
       });
+    },
+    [ReturnFormat.OBJECT]: (computedStyle) => {
+      const changes: { [key: string]: CSSPropertyInfo } = {};
 
-      // Invoke callback only if there are changes
-      if (Object.keys(changedProperties).length > 0) {
-        this._callback(changedProperties);
-      }
+      this._processChanges(computedStyle, (propertyName, currentValue, previousValue, hasChanged) => {
+        changes[propertyName] = {
+          propertyName,
+          value: currentValue,
+          previousValue,
+          changed: hasChanged,
+        };
+      }, () => {
+        if (Object.keys(changes).length > 0) {
+          this._callback(changes);
+        }
```

Comment:
This is duplicated in both handlers that are currently supported.

I’d rather see this not duplicated. Maybe an idea here is that a handler should simply return its results, and then have this calling of the callback be part of the `_handleUpdate` method?

Reply:
I’ve de-duplicated the callback invocation and it’s now part of the `_handleUpdate` method